src/Features/LanguageServer/Protocol/Handler/MapCode/MapCodeHandler.cs

```diff
@@ -0,0 +1,152 @@
+
+        using var _ = PooledDictionary<Uri, TextEdit[]>.GetInstance(out var uriToEditsMap);
+        foreach (var codeMapping in request.Mappings)
+        {
+            var mappingResult = await MapCodeAsync(codeMapping).ConfigureAwait(false);
+
+            if (mappingResult is not (Uri uri, TextEdit[] textEdits))
+            {
+                // Failed the entire request if any of the sub-requests failed
+                context.TraceWarning("mapCode Request failed: a sub-request failed");
+                return null;
+            }
+
+            if (!uriToEditsMap.TryAdd(uri, textEdits))
+                throw new ArgumentException($"mapCode sub-request for {uri} failed: multiple MapCodeMappings for the same document is not supported.");
+        }
+
+        // return a combined WorkspaceEdit
+        if (context.GetRequiredClientCapabilities().Workspace?.WorkspaceEdit?.DocumentChanges is true)
+        {
+            return new WorkspaceEdit
+            {
+                DocumentChanges = uriToEditsMap.Select(kvp => new TextDocumentEdit
+                {
+                    TextDocument = new OptionalVersionedTextDocumentIdentifier { Uri = kvp.Key },
+                    Edits = kvp.Value,
+                }).ToArray()
+            };
+        }
+        else
+        {
+            return new WorkspaceEdit
+            {
+                Changes = uriToEditsMap.ToDictionary(kvp => ProtocolConversions.GetDocumentFilePathFromUri(kvp.Key), kvp => kvp.Value)
+            };
+        }
+
+        async Task<(Uri, TextEdit[])?> MapCodeAsync(LSP.MapCodeMapping codeMapping)
+        {
+            var textDocument = codeMapping.TextDocument
+                ?? throw new ArgumentException($"mapCode sub-request failed: MapCodeMapping.TextDocument not expected to be null.");
+
+            if (context.Solution.GetDocument(textDocument) is not Document document)
+                throw new ArgumentException($"mapCode sub-request for {textDocument.Uri} failed: can't find this document in the workspace.");
+
+            var codeMapper = document.GetLanguageService<IMapCodeService>()
+                ?? throw new ArgumentException($"mapCode sub-request for {textDocument.Uri} failed: No IMapCodeService for this document.");
+
+            var focusLocations = await ConvertFocusLocationsToDocumentAndSpansAsync(
+                document,
+                textDocument,
+                codeMapping.FocusLocations,
+                cancellationToken).ConfigureAwait(false);
+
+            var newDocument = await codeMapper.MapCodeAsync(
+                document,
+                codeMapping.Contents.ToImmutableArrayOrEmpty(),
+                focusLocations,
+                formatMappedCode: false,
+                cancellationToken).ConfigureAwait(false);
+
+            if (newDocument is null)
+            {
+                context.TraceWarning($"mapCode sub-request for {textDocument.Uri} failed: 'IMapCodeService.MapCodeAsync' returns null.");
```

Comment:
is this a failure?  shouldn't it be normal to not be able to map code?  issuing a warning indicates something teh user should be concerned about right?  can this just be info?